### PR TITLE
Add slow SQL query warning

### PR DESCRIPTION
**Pull request by earl-warren**
- Databases are one of the most important parts of Forgejo, every interaction uses the database in one way or another. Therefore, it is important to maintain the database and recognize when the server is not doing well with the database. There already is the option to log *every* SQL query along with its execution time, but monitoring becomes impractical for larger instances and takes up unnecessary storage in the logs.
- Add a QoL enhancement that allows instance administrators to specify a threshold value beyond which query execution time is logged as a warning in the xorm logger. The default value is a conservative five seconds to avoid this becoming a source of spam in the logs.
- The use case for this patch is that with an instance the size of Codeberg, monitoring SQL logs is not very fruitful and most of them are uninteresting. Recently, in the context of persistent deadlock issues (https://codeberg.org/forgejo/forgejo/issues/220), I have noticed that certain queries hold locks on tables like comment and issue for several seconds. This patch helps to identify which queries these are and when they happen.
- Added unit test.

(cherry picked from commit 9cf501f1af4cd870221cef6af489618785b71186)

### CODE REVIEW

**Comment by JakobDev** [comment]
The new Setting needs to be documented

File: models/db/engine_test.go [hunk 1]
```diff
@@ -84,3 +88,37 @@ func TestPrimaryKeys(t *testing.T) {
 		}
 	}
 }
+
+func TestSlowQuery(t *testing.T) {
```
Comment (wxiaoguang) [inline models/db/engine_test.go]: The test is not right.

I know you like long discussions like https://codeberg.org/forgejo/discussions/issues/54 (for testing), but I think you should first understand the code correctly and write tests correctly.

File: models/db/engine.go [hunk 2]
```diff
@@ -299,3 +309,21 @@ func SetLogSQL(ctx context.Context, on bool) {
...
+
+var _ contexts.Hook = &SlowQueryHook{}
+
+func (SlowQueryHook) BeforeProcess(c *contexts.ContextHook) (context.Context, error) {
+	return c.Ctx, nil
+}
+
+func (h *SlowQueryHook) AfterProcess(c *contexts.ContextHook) error {
+	if c.ExecuteTime >= h.Treshold {
+		h.Logger.Log(8, log.WARN, "[Slow SQL Query] %s %v - %v", c.SQL, c.Args, c.ExecuteTime)
```
Comment (delvh) [inline models/db/engine.go:328]: Why `8`?
What makes you so certain that it must be `8`?
If you are certain that `8` is correct, document **why**?

File: models/db/engine_test.go (hunk 1 above)
Comment (delvh) [reply models/db/engine_test.go]: I do agree with @wxiaoguang, this test looks really weird, and not like it tests what it is supposed to test.
I don't even know what the intention of this test is, the implementation looks straightforward enough for me not to warrant a test.
There are far worse things without tests that should have a test.

Comment (earl-warren) [reply models/db/engine_test.go]: Removed.

**Review by delvh — APPROVED** [review]

**Review by denyskon — APPROVED** [review]

File: custom/conf/app.example.ini [hunk 3]
```diff
@@ -412,6 +412,10 @@ USER = root
 ;;
 ;; Whether execute database models migrations automatically
 ;AUTO_MIGRATION = true
+;;
+;; Threshold value (in seconds) beyond which query execution time is logged as a warning in the xorm logger
+;;
+;SLOW_QUERY_TRESHOLD = 5s
```
Comment (wxiaoguang) [inline custom/conf/app.example.ini]: Well, I think it is a typo ..... and it needs to be fixed at least before merging.

`THRESHOLD`

**Comment by wxiaoguang** [comment]
More thoughts about this feature:

1. Is it useful? Somewhat, but not really. DBA should already have the ability to find the slow queries from database log.
2. Is it complete? I think no, because there are also many performance problems are caused "plenty SQLs in a loop", say, even if a single SQL is fast, if it is executed in a loop, the whole request will slow down. But this PR isn't able to catch this case.

So I think it is fine to merge, although it is far from ideal.

ps: my previous about the "tests", I never meant that it doesn't need tests, but actually that test code was wrong, so I think the tests should be rewritten correctly.

**Review by 6543 — APPROVED** [review]
i think it's a usefull tool ...
... yes it does not cover the whole picture, but if it catches at lest some, its better than nothing

**Pull request merged**